waku/v2/api/publish/rln_rate_limiting.go

```diff
@@ -0,0 +1,64 @@
+package publish
+
+import (
+	"context"
+	"errors"
+	"sync"
+	"time"
+
+	"go.uber.org/zap"
+)
+
+var ErrRateLimited = errors.New("rate limit exceeded")
+
+const RlnLimiterCapacity = 100
+const RlnLimiterRefillInterval = 10 * time.Minute
+
+// RlnRateLimiter is used to rate limit the outgoing messages,
+// The capacity and refillInterval comes from RLN contract configuration.
+type RlnRateLimiter struct {
+	mu             sync.Mutex
+	capacity       int
+	tokens         int
+	refillInterval time.Duration
+	lastRefill     time.Time
+}
+
+// NewRlnPublishRateLimiter creates a new rate limiter, starts with a full capacity bucket.
+func NewRlnRateLimiter(capacity int, refillInterval time.Duration) *RlnRateLimiter {
+	return &RlnRateLimiter{
+		capacity:       capacity,
+		tokens:         capacity, // Start with a full bucket
+		refillInterval: refillInterval,
+		lastRefill:     time.Now(),
+	}
+}
+
+// Allow checks if a token can be consumed, and refills the bucket if necessary
+func (rl *RlnRateLimiter) Allow() bool {
+	rl.mu.Lock()
+	defer rl.mu.Unlock()
+
+	// Refill tokens if the refill interval has passed
+	now := time.Now()
+	if now.Sub(rl.lastRefill) >= rl.refillInterval {
+		rl.tokens = rl.capacity // Refill the bucket
+		rl.lastRefill = now
+	}
+
+	// Check if there are tokens available
+	if rl.tokens > 0 {
+		rl.tokens--
+		return true
+	}
+
+	return false
+}
+
+func (rl *RlnRateLimiter) Check(ctx context.Context, logger *zap.Logger) error {
+	if rl.Allow() {
+		return nil
+	}
+	logger.Error("could not send message rate limited", zap.Error(ErrRateLimited))
```

Comment:
Consider not logging this as there's already a log in https://github.com/status-im/status-go/blob/develop/wakuv2/message_publishing.go#L104 for when message fails to publish.
You would end up with:
```
2024-12-05T21:11:22.668Z    ERROR  somefile.go:1547  could not send message rate limited {"error": "rate limit exceeded"}
2024-12-05T21:11:22.668Z    ERROR  somefile.go:1547  could not send message {"envelopeHash": "0xSomeHash", "pubsubTopic": "someTopic", "contentTopic": "someContentTopic", "timestamp": 1234567890, "error": "rate limit exceeded"}
```